lib/spack/llnl/util/filesystem.py

```diff
@@ -1816,66 +1823,67 @@ def find(root, files, recursive=True):
 
     Returns:
         list: The files that have been found
     """
     if isinstance(files, str):
         files = [files]
 
-    if recursive:
-        tty.debug(f"Find (recursive): {root} {str(files)}")
-        result = _find_recursive(root, files)
-    else:
-        tty.debug(f"Find (not recursive): {root} {str(files)}")
-        result = _find_non_recursive(root, files)
+    if max_depth is not _unset and not recursive:
+        raise ValueError(f"max_depth ({max_depth}) cannot be set if recursive is False")
+
+    if not recursive:
+        max_depth = 0
+
+    tty.debug(f"Find (max depth = {max_depth}): {root} {str(files)}")
+    result = find_max_depth(root, files, max_depth)
 
     tty.debug(f"Find complete: {root} {str(files)}")
     return result
 
 
-@system_path_filter
-def _find_recursive(root, search_files):
-    # The variable here is **on purpose** a defaultdict. The idea is that
-    # we want to poke the filesystem as little as possible, but still maintain
-    # stability in the order of the answer. Thus we are recording each library
-    # found in a key, and reconstructing the stable order later.
-    found_files = collections.defaultdict(list)
-
-    # Make the path absolute to have os.walk also return an absolute path
-    root = os.path.abspath(root)
-    for path, _, list_files in os.walk(root):
-        for search_file in search_files:
-            matches = glob.glob(os.path.join(path, search_file))
-            matches = [os.path.join(path, x) for x in matches]
-            found_files[search_file].extend(matches)
+@system_path_filter(arg_slice=slice(1))
+def find_max_depth(root, globs, max_depth=_unset):
+    """Given a set of non-recursive glob file patterns, finds all
+    files matching those patterns up to a maximum specified depth.
 
-    answer = []
-    for search_file in search_files:
-        answer.extend(found_files[search_file])
+    Does not search in directories below the specified depth.
 
-    return answer
+    If ``globs`` is a list, files matching earlier entries are placed
+    in the return value before files matching later entries.
+    """
+    if not os.path.exists(root):
+        return []
 
+    if isinstance(globs, str):
+        globs = [globs]
 
-@system_path_filter
-def _find_non_recursive(root, search_files):
-    # The variable here is **on purpose** a defaultdict as os.list_dir
-    # can return files in any order (does not preserve stability)
     found_files = collections.defaultdict(list)
 
-    # Make the path absolute to have absolute path returned
-    root = os.path.abspath(root)
+    dir_queue = collections.deque([(0, root)])
+    while dir_queue:
+        depth, next_dir = dir_queue.pop()
+        if (max_depth is _unset) or depth < max_depth:
+            for dir_entry in os.scandir(next_dir):
+                if dir_entry.is_dir():
+                    dir_path = os.path.join(next_dir, dir_entry.name)
+                    if not os.path.islink(dir_path):
```

Comment:
Don't do another stat call, the point of the `os.scandir` iterator is that the stat call can be avoided.

`dir_entry.is_dir(follow_symlinks=False)`.